File: webnotes/model/doclist.py
# Copyright (c) 2013, Web Notes Technologies Pvt. Ltd.
# MIT License. See license.txt 

import webnotes
import webnotes.model
from webnotes.model.doc import Document

class DocList(list):
	"""DocList object as a wrapper around a list"""
	def get(self, filters, limit=0):
		"""pass filters as:
			{"key": "val", "key": ["!=", "val"],
			"key": ["in", "val"], "key": ["not in", "val"], "key": "^val",
			"key" : True (exists), "key": False (does not exist) }"""

		out = []
		
		for doc in self:
			d = isinstance(getattr(doc, "fields", None), dict) and doc.fields or doc
			add = True
			for f in filters:
				fval = filters[f]
				
				if fval is True:
					fval = ["not None", fval]
				elif fval is False:
					fval = ["None", fval]
				elif not isinstance(fval, list):
					if isinstance(fval, basestring) and fval.startswith("^"):
						fval = ["^", fval[1:]]
					else:
						fval = ["=", fval]
				
				if not webnotes.compare(d.get(f), fval[0], fval[1]):
					add = False
					break

			if add:
				out.append(doc)
				if limit and (len(out)-1)==limit:
					break
		
		return DocList(out)
		
	def get_distinct_values(self, fieldname):
		return list(set(map(lambda d: d.fields.get(fieldname), self)))

	def remove_items(self, filters):
		for d in self.get(filters):
			self.remove(d)

	def getone(self, filters):
		return self.get(filters, limit=1)[0]

	def copy(self):
		out = []
		for d in self:
			if isinstance(d, dict):
				fielddata = d
			else:
				fielddata = d.fields
			fielddata.update({"name": None})
			out.append(Document(fielddata=fielddata))
		return DocList(out)
		
	def get_item_value(self, d, name):
		if isinstance(d, dict):
			return d.get(name)
		else:
			return d.fields.get(name)
		
	def filter_valid_fields(self):
		import webnotes.model
		fieldnames = {}
		for d in self:
			remove = []
			for f in d:
				if f not in fieldnames.setdefault(d.doctype,
						webnotes.model.get_fieldnames(d.doctype)):
					remove.append(f)
			for f in remove:
				del d[f]
				
	def append(self, doc):
		if not isinstance(doc, Document):
			doc = Document(fielddata=doc)
			
		self._prepare_doc(doc)

		super(DocList, self).append(doc)
		
	def extend(self, doclist):
		doclist = objectify(doclist)
		for doc in doclist:
			self._prepare_doc(doc)
		
		super(DocList, self).extend(doclist)
		
		return self
		
	def _prepare_doc(self, doc):
		if not doc.name:
			doc.fields["__islocal"] = 1
			doc.docstatus = 0
		if doc.parentfield:
			if not doc.parenttype:
				doc.parenttype = self[0].doctype
			if not doc.parent:
				doc.parent = self[0].name
			if not doc.idx:
				siblings = [int(self.get_item_value(d, "idx") or 0) for d in self.get({"parentfield": doc.parentfield})]
				doc.idx = (max(siblings) + 1) if siblings else 1
		
def objectify(doclist):
	from webnotes.model.doc import Document
	return map(lambda d: isinstance(d, Document) and d or Document(d), doclist)

File: public/js/wn/model/doclist.js
// Copyright (c) 2013, Web Notes Technologies Pvt. Ltd. and Contributors
// MIT License. See license.txt 

wn.provide("wn.model");
wn.model.DocList = Class.extend({
	init: function(doctype, name) {
		this.doctype = doctype; this.name = name;
		this.doclist = wn.model.get_doclist(this.doctype, this.name);
		this.doc = this.doclist[0];
	},
	
	save: function(action, callback, btn) {
		this.check_name();
		if(this.check_mandatory()) {
			var me = this;
			this._call({
				method: "webnotes.widgets.form.save.savedocs",
				args: { docs: wn.model.compress(this.doclist), action:action},
				callback: function(r) {
					$(document).trigger("save", me.doc);
					callback(r);
				},
				btn: btn
			});
		}
	},
	
	cancel: function(callback, btn) {
		var me = this;
		this._call({
			method: "webnotes.widgets.form.save.cancel",
			args: { doctype: this.doctype, name: this.name },
			callback: function(r) {
				$(document).trigger("save", wn.model.get_doc(me.doctype, me.name));
				callback(r);
			},
			btn: btn
		});
	},
	
	check_name: function() {
		var doc = this.doclist[0];
		var meta = locals.DocType[doc.doctype];
		if(doc.__islocal && (meta && meta.autoname 
				&& meta.autoname.toLowerCase()=='prompt')) {
			var newname = prompt('Enter the name of the new '+ doc.doctype, '');
			if(newname) { 
				doc.__newname = strip(newname);
			} else {
				msgprint("Name is required.");
				throw "name required";
			}
		}
	},
	
	check_mandatory: function() {
		var me = this;
		var has_errors = false;
		this.scroll_set = false;
		
		if(this.doc.docstatus==2) return true; // don't check for cancel
		
		$.each(this.doclist, function(i, doc) {
			
			var error_fields = [];
			
			$.each(wn.meta.docfield_list[doc.doctype] || [], function(i, docfield) {
				if(docfield.fieldname) {
					var df = wn.meta.get_docfield(doc.doctype, 
						docfield.fieldname, me.doclist[0].name);

					if(df.reqd && !wn.model.has_value(doc.doctype, doc.name, df.fieldname)) {
						has_errors = true;
						error_fields[error_fields.length] = df.label;
						
						// scroll to field
						if(!me.scroll_set) 
							me.scroll_to(doc.parentfield || df.fieldname);
					}
					
				}
			});
			if(error_fields.length)
				msgprint('<b>Mandatory fields required in '+ (doc.parenttype 
					? (wn.meta.docfield_map[doc.parenttype][doc.parentfield].label + ' (Table)') 
					: doc.doctype) + ':</b>\n' + error_fields.join('\n'));	
		});
		
		return !has_errors;
	},
	
	scroll_to: function(fieldname) {
		var f = cur_frm.fields_dict[fieldname];
		if(f) {
			$(document).scrollTop($(f.wrapper).offset().top - 100);
		}
		this.scroll_set = true;
	},

	_call: function(opts) {
		// opts = {
		// 	method: "some server method",
		// 	args: {args to be passed},
		// 	callback: callback,
		// 	btn: btn
		// }
		$(opts.btn).prop("disabled", true);
		return wn.call({
			freeze: true,
			method: opts.method,
			args: opts.args,
			callback: function(r) {
				$(opts.btn).prop("disabled", false);
				opts.callback && opts.callback(r);
			}
		})
	},
});